plugin/pkg/auth/authenticator/token/keystone/keystone.go

```diff
@@ -0,0 +1,119 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+	"io/ioutil"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+	"k8s.io/kubernetes/pkg/auth/user"
+	openstackcp "k8s.io/kubernetes/pkg/cloudprovider/providers/openstack"
+)
+
+// KeystoneAuthenticator contacts openstack keystone to validate user's credentials passed in the request.
+// The keystone endpoint is passed during apiserver startup
+type KeystoneAuthenticator struct {
+	configPath string
+}
+
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	glog.V(3).Infof("Starting Keystone Token Auth.")
+	cfg, provider, err := LoadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	glog.V(3).Infof("Getting OpenStack Identity Provider...")
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct {
+		User struct { Id string }
+		Project struct{ Id string }
+		Roles []struct{ Name string }
+	}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	extra := make(map[string][]string, 3)
+	extra["kubernetes.io/keystone_authn"] = []string{ "true" }
+	if obj.Token.Project.Id != "" {
+		extra["kubernetes.io/keystone_project_id"] = []string{ obj.Token.Project.Id }
+	}
+	if obj.Token.Roles != nil && len(obj.Token.Roles) > 0 {
+		l := make([]string, len(obj.Token.Roles))
+		for i:= 0; i < len(obj.Token.Roles); i++ {
+			l[i] = obj.Token.Roles[i].Name
+		}
+		extra["kubernetes.io/keystone_roles"] = l
+	}
+	return &user.DefaultInfo{ Name: obj.Token.User.Id, Extra:extra }, response.StatusCode == 200, nil
```

Comment:
Yeah, I think this is a greater conversation with the SigAuth team. username is not unique in keystone. neither are groups. So those probably should not be used directly if any code anywhere in kubernetes assumes they are unique.
userid in keystone's the only unique string for a user.